mirage/stub/dns_stub_mirage.ml

```diff
@@ -0,0 +1,401 @@
+      | `Tcp_connections -> "tcp-connections"
+      | `Recursive_queries -> "recursive-questions"
+      | `Recursive_answers -> "recursive-answers"
+      | `Recursive_connections -> "recursive-connections"
+      | `Authoritative_answers -> "authoritative-answers"
+      | `Authoritative_errors -> "authoritative-errors"
+      | `Reserved_answers -> "reserved-answers"
+      | `On_update -> "on-update"
+      | `Resolver_queries -> "resolver-queries"
+      | `Resolver_answers -> "resolver-answers"
+      | `Resolver_nodata -> "resolver-nodata"
+      | `Resolver_nodomain -> "resolver-nodomain"
+      | `Resolver_servfail -> "resolver-servfail"
+      | `Resolver_notimp -> "resolver-notimplemented"
+    in
+    let metrics = Dns.counter_metrics ~f "stub-resolver" in
+    (fun x -> Metrics.add metrics (fun x -> x) (fun d -> d x))
+
+  module IM = Map.Make(struct
+      type t = int
+      let compare : int -> int -> int = fun a b -> compare a b
+    end)
+
+  module Client = struct
+    module Transport : Dns_client.S
+      with type stack = S.t
+       and type +'a io = 'a Lwt.t
+       and type io_addr = Ipaddr.V4.t * int = struct
+      type stack = S.t
+      type io_addr = Ipaddr.V4.t * int
+      type ns_addr = [`TCP | `UDP] * io_addr
+      type +'a io = 'a Lwt.t
+
+      type t = {
+        rng : (int -> Cstruct.t) ;
+        nameserver : ns_addr ;
+        stack : stack ;
+        mutable flow : S.TCPV4.flow option ;
+        mutable requests : (Cstruct.t, [ `Msg of string ]) result Lwt_condition.t IM.t ;
+      }
+      type flow = { t : t ; mutable id : int }
+
+      let create
+          ?rng
+          ?(nameserver = `TCP, (Ipaddr.V4.of_string_exn "91.239.100.100", 53))
+          stack =
+        let rng = match rng with None -> R.generate ?g:None | Some x -> x in
+        { rng ; nameserver ; stack ; flow = None ; requests = IM.empty }
+
+      let nameserver { nameserver ; _ } = nameserver
+      let rng { rng ; _ } = rng
+
+      let bind = Lwt.bind
+      let lift = Lwt.return
+
+      let cancel_all t =
+        IM.iter (fun _id cond ->
+            Lwt_condition.broadcast cond (Error (`Msg "disconnected")))
+          t.requests
+
+      let rec read_loop ?(linger = Cstruct.empty) t flow =
+        S.TCPV4.read flow >>= function
+        | Error e ->
+          t.flow <- None;
+          Log.err (fun m -> m "error %a reading from resolver" S.TCPV4.pp_error e);
+          cancel_all t;
+          Lwt.return_unit
+        | Ok `Eof ->
+          t.flow <- None;
+          Log.info (fun m -> m "end of file reading from resolver");
+          cancel_all t;
+          Lwt.return_unit
+        | Ok (`Data cs) ->
+          let rec handle_data data =
+            let cs_len = Cstruct.len data in
+            if cs_len > 2 then
+              let len = Cstruct.BE.get_uint16 data 0 in
+              if cs_len - 2 >= len then
+                let packet, rest = Cstruct.split data (len + 2) in
```

Comment:
this will always allocate two new buffers, would it make sense to have something like
```ocaml
let packet, rest =
  if cs_len - 2 = len then packet, Cstruct.empty
  else Cstruct.split data (len + 2)
in
```
to prevent the extra allocations in the case where there's no trailing data? That does mean that we're passing on a mutable reference to `packet`, but as far as I can tell we don't use `data` again in that case.